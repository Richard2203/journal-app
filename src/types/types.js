export const types = {
	login: '[Auth] Login',
	logout: '[Auth] Logout',

	uiSetError: '[UI] ErrorFormValues',
	uiRemoveError: '[UI] ErrorRemoveFormValues',
};
