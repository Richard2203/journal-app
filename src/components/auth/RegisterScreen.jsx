import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { removeError, setError } from '../../action/ui';

const RegisterScreen = () => {
	// el custom hook useDispatch proviene de redux y
	// sirve para retornar una funcion que recibe un metodo mediante el cual
	// tendremos acceso a Dispatch enviandole informacion
	const dispatch = useDispatch();

	const [formValue, hadleInputChange] = useForm({
		name: '',
		email: '',
		password: '',
		passwordConfirm: '',
	});

	const { name, email, password, passwordConfirm } = formValue;

	const handleRegister = (e) => {
		e.preventDefault();
		if (isFormValid()) {
			console.log(name, email, password, passwordConfirm);
		}
	};

	const isFormValid = () => {
		if (validator.isEmpty(name)) {
			dispatch(setError('The name canÂ´t null'));
			return false;
		} else if (!validator.isEmail(email)) {
			dispatch(setError('The email is not valid'));
			return false;
		} else if (!(password === passwordConfirm)) {
			dispatch(setError('The passwords not equals'));
			return false;
		} else if (!validator.isStrongPassword(password)) {
			dispatch(setError('The password is not strong'));
			return false;
		}
		dispatch(removeError());
		return true;
	};

	return (
		<>
			<h3 className="auth__title">Register</h3>
			<div className="Form__valid-alert">
				<span>Hola</span>
			</div>

			<form onSubmit={handleRegister}>
				<div className="form-control">
					<input
						type="text"
						id="name"
						name="name"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={name}
						onChange={hadleInputChange}
					/>

					<label htmlFor="name" className="form-control__label">
						Name
					</label>
				</div>

				<div className="form-control">
					<input
						type="text"
						id="email"
						name="email"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={email}
						onChange={hadleInputChange}
					/>

					<label htmlFor="email" className="form-control__label">
						Email
					</label>
				</div>

				<div className="form-control">
					<input
						type="password"
						id="password"
						name="password"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={password}
						onChange={hadleInputChange}
					/>

					<label htmlFor="password" className="form-control__label">
						Password
					</label>
				</div>

				<div className="form-control">
					<input
						type="password"
						id="passwordConfirm"
						name="passwordConfirm"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={passwordConfirm}
						onChange={hadleInputChange}
					/>

					<label htmlFor="password" className="form-control__label">
						Confirm password
					</label>
				</div>

				<button type="submit" className="btn cursor btn-block">
					Login
				</button>
			</form>

			<hr />

			<Link to="/auth/login" className="link">
				Already Register?
			</Link>
		</>
	);
};

export default RegisterScreen;
