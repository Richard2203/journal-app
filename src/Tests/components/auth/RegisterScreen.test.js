// Instalando la dependencia para poder hacer pruebas mock en store de redux
//  https://www.npmjs.com/package/redux-mock-store
// se añaden las importaciones
//  import configureStore from 'redux-mock-store';
//  const middlewares = [thunk];
//  const mockStore = configureStore(middlewares);
//  const store = mockStore({});

import configureStore from 'redux-mock-store';
//importamos el middleware que usaremos para pruebas
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import RegisterScreen from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
	auth: {},
	ui: { msgError: null, loading: false },
};

// mocksStore recibe objetos para establecer el store; en este caso
// el store contiene el objeto: auth
let store = mockStore(initState);

// - Debido a que LoginScreen emplea Redux debemos pasarle la creacion
//   de los stores mediante Provider y su propieadad store
// - Debido a que se emplea la etiqueta Link se deben pasar las referencias
//   a este componente mediante MemoryRouter, el cual hara de simulador de esta
//   funcion
const wrapper = mount(
	<Provider store={store}>
		<MemoryRouter>
			<RegisterScreen />
		</MemoryRouter>
	</Provider>
);

describe('Pruebas en el componente RegisterScreen', () => {
	// Para poder realizar pruebas en redux store se emplean mocks
	//      npm install redux-mock-store --save-dev

	test('debe de renderizarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe de ejecutar el dispatch', () => {
		// const inputName = wrapper.find('input[name="name"]');

		// este simulate es para establecer el campo name en valor '', sin embargo
		// por defecto ya tiene ese valor asi que no es necesario
		// inputName.simulate('change', {
		// 	target: {
		// 		value: '',
		// 		name: 'name',
		// 	},
		// });

		wrapper.find('form').simulate('submit', { preventDefault() {} });

		const actions = store.getActions();

		expect(actions[0]).toEqual({
			type: types.uiSetError,
			payload: 'The name can´t null',
		});
	});
});
