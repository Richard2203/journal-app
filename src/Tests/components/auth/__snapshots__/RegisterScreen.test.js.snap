// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente RegisterScreen debe de renderizarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="name"
        name="name"
        onChange={[Function]}
        placeholder=" "
        type="text"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="name"
      >
        Name
      </label>
    </div>
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder=" "
        type="text"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="email"
      >
        Email
      </label>
    </div>
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="password"
      >
        Password
      </label>
    </div>
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="passwordConfirm"
        name="passwordConfirm"
        onChange={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="password"
      >
        Confirm password
      </label>
    </div>
    <button
      className="btn cursor btn-block"
      type="submit"
    >
      Login
    </button>
  </form>,
  <hr />,
  <a
    className="link"
    href="/auth/login"
    onClick={[Function]}
  >
    Already Register?
  </a>,
]
`;
